File: medium/978.longest-turbulent-subarray.java
/*
 * @lc app=leetcode id=978 lang=java
 *
 * [978] Longest Turbulent Subarray
 */

// @lc code=start
class Solution {
    public int maxTurbulenceSize(int[] arr) {
        // dp(i) = max turbulence size for arr[0...i]
        // dp(0) = 1
        // dp(1) = 2
        // dp(2) = if a[1] > a[0] => { if a[2] < a[1] => dp(1) + 1, else => 1}
        //       = if a[1] < a[0] => { if a[2] > a[1] => dp(1) + 1, else => 1}
        // dp(i) = if a[i-1] > a[i-2] => { if a[i] < a[i-1] => dp(i-1) + 1, else => 2}
        //       = if a[i-1] < a[i-2] => { if a[i] > a[i-1] => dp(i-1) + 1, else => 2}
        // return max(dp(i))
        if (arr.length <= 1) {
            return arr.length;
        }
        

        int dp1 = 0;
        if (arr[0] == arr[1]) {
            dp1 = 1;
        } else {
            dp1 = 2;
        }
        
        
        int ans = dp1;
        int dp = 0;
        for (int i=2; i<arr.length; i++) {
            if (dp1 == 1) {
                if (arr[i] == arr[i-1]) {
                    dp = 1;
                } else {
                    dp = 2;
                }
            } else {
                if (arr[i-1] > arr[i-2]) {
                    if (arr[i] < arr[i-1]) {
                        dp = dp1 + 1;
                    } else if (arr[i] == arr[i-1]) {
                        dp = 1;
                    } else {
                        dp = 2;
                    }
                }

                if (arr[i-1] < arr[i-2]) {
                    if (arr[i] > arr[i-1]) {
                        dp = dp1 + 1;
                    } else if (arr[i] == arr[i-1]) {
                        dp = 1;
                    } else {
                        dp = 2;
                    }
                }
            }

            //System.out.println("idx: " + i + " dp: " + dp);
            ans = Math.max(ans, dp);
            dp1 = dp;
        }
        
        return ans;
    }

    private int myFirstSolution(int[] arr) {
        if (arr.length <= 1) {
            return arr.length;
        }

        int l = 0;
        int ans = 0;
        int r = 1;
        while (r<arr.length) {
            if (!check(arr, l, r)) {
                // update answer
                ans = Math.max(ans,  r - l);
                if (arr[r] == arr[l]) {
                    if (r == arr.length - 1) {
                        break;
                    } else {
                        l = r;
                    }
                } else {
                    l = r - 1;
                }
            } else if (r == arr.length - 1) {
                ans = Math.max(ans,  arr.length - l);
            }

            r++;
        }

        // [0,8,45,88,48,68,28,55,17,24]
        // [9, 9]
        // [100,100,100]
        return ans;
    }

    private boolean check(int[] arr, int l, int r) {
        int lastSign = 0; // 1 => arr[i] > arr[i-1] -1 => arr[i] < arr[i-1]
        for (int i=l; i<=r; i++) {
            if (i > l) {
                if (arr[i] == arr[i-1]) {
                    return false;
                }
                if ((arr[i] > arr[i-1] && lastSign == 1) || (arr[i] < arr[i-1] && lastSign == -1)) {
                    return false;
                } else {
                    lastSign = arr[i] > arr[i-1] ? 1 : -1;
                }
            }
        }



        return true;
    }
}
// @lc code=end

File: medium/978.longest-turbulent-subarray.js
/*
 * @lc app=leetcode id=978 lang=javascript
 *
 * [978] Longest Turbulent Subarray
 * [dp]
 */

// @lc code=start
/**
 * @param {number[]} arr
 * @return {number}
 */
var maxTurbulenceSize = function(arr) {
  // in a common case, we should use dp(i) for the LTS of arr[0...i]
  // but in this case, only one array dp(i) is not enough
  // we need to also remember if the previous pair of numbers go up or they go down.
  // so we will have two dp arrays:
  // up(i): LTS of arr[0...i] and arr[i] > arr[i-1], the last two numbers go up
  // down(i): LTS of arr[0...i] and arr[i] < arr[i-1], the last two numbers go down
  // initialize up(i) = down(i) = 1, since at least the len is 1
  // if arr[i] > arr[i-1]: up(i) = down(i-1) + 1
  // if arr[i] < arr[i-1]: down(i) = up(i-1) + 1

  // finally, return max(up, down)
  if (arr.length < 2) {
    return arr.length;
  }

  let ans = 1;
  const up = new Array(arr.length).fill(1);
  const down = new Array(arr.length).fill(1);
  for (let i=1; i<arr.length; i++) {
    if (arr[i] > arr[i-1]) {
      up[i] = down[i-1] + 1;
    } else if (arr[i] < arr[i-1]) {
      down[i] = up[i-1] + 1;
    }

    ans = Math.max(ans, Math.max(up[i], down[i]));
  }

  return ans;
};
// @lc code=end

// wrong solution:
// we can define dp[i] as the longest turbulent sub array arr[0...i]
// so we have the following
// dp(0) = 1;
// dp(1) = 2 if arr[1] != arr[0] else 1
// for dp(i), we need to check three elements, arr[i], arr[i-1], arr[i-2]
// case1. if arr[i-1] > arr[i-2]
//          if arr[i] < a[i-1] => dp(i-1) + 1
//          else => dp(i-1) 

// case2. if arr[i-1] < arr[i-2]
//          if arr[i] > arr[i-1] => dp(i-1) + 1
//          else => dp(i-1)

// case3. if arr[i-1] = arr[i-2]: dp(i) = max(dp(i-1), 1) = dp(i-1);

// return max(dp)